src/Init/Data/BitVec/Bitblast.lean

```diff
@@ -429,6 +429,417 @@ theorem shiftLeft_eq_shiftLeftRec (x : BitVec w₁) (y : BitVec w₂) :
+    have := h.hwrn
+    omega
+
+/--
+This is used when proving the correctness of the divison algorithm,
+where we know that `r < d`.
+We then want to show that `((r.shiftConcat b) - d) < d` as the loop invariant.
+In arithmethic, this is the same as showing that
+`r * 2 + 1 - d < d`, which this theorem establishes.
+-/
+private theorem two_mul_add_sub_lt_of_lt_of_lt_two (h : a < x) (hy : y < 2) :
+    2 * a + y - x < x := by omega
+
+/-- We show that the output of `divSubtractShift` is lawful, which tells us that it
+obeys the division equation. -/
+def divSubtractShiftProof (qr : DivModState w) (h : qr.LawfulShiftSubtract  wr wn n d) :
+    DivModState.Lawful w (wr + 1) (wn - 1) n d (divSubtractShift n d wn qr) := by
+  simp only [divSubtractShift, decide_eq_true_eq]
+  -- We add these hypotheses for `omega` to find them later.
+  have ⟨⟨hrwn, hd, hrd, hr, hn, hrnd⟩, hwn_lt⟩ := h
+  have : d.toNat * (qr.q.toNat * 2) = d.toNat * qr.q.toNat * 2 := by rw [Nat.mul_assoc]
+  by_cases rltd : shiftConcat qr.r (n.getLsb (wn - 1)) < d
+  · simp only [rltd, ↓reduceIte]
+    constructor <;> try bv_omega
+    case pos.hrWidth => apply toNat_shiftConcat_lt_of_lt_of_lt_two_pow <;> omega
+    case pos.hqWidth => apply toNat_shiftConcat_lt_of_lt_of_lt_two_pow <;> omega
+    case pos.hdiv =>
+      simp [qr.toNat_shiftRight_sub_one_eq h,
+      toNat_shiftConcat_eq_of_lt_of_lt_two_pow
+        (hk := qr.wr_lt_w h)
+        (hx := h.hrWidth),
+      h.hdiv,
+      toNat_shiftConcat_eq_of_lt_of_lt_two_pow (qr.wr_lt_w h) h.hqWidth, Nat.add_mul, toNat_false, Nat.add_zero,
+      this]
+      omega
+  · simp only [rltd, ↓reduceIte]
+    constructor <;> try bv_omega
+    case neg.hrLtDivisor =>
+      simp only [lt_def, Nat.not_lt] at rltd
+      rw [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le rltd,
+        toNat_shiftConcat_eq_of_lt_of_lt_two_pow (hk := qr.wr_lt_w h) (hx := h.hrWidth),
+        Nat.mul_comm]
+      apply two_mul_add_sub_lt_of_lt_of_lt_two <;> bv_omega
+    case neg.hrWidth =>
+      simp only
+      have hdr' : d ≤ (qr.r.shiftConcat (n.getLsb (wn - 1))) :=
+        BitVec.le_iff_not_lt.mp rltd
+      have hr' : ((qr.r.shiftConcat (n.getLsb (wn - 1)))).toNat < 2 ^ (wr + 1) := by
+        apply toNat_shiftConcat_lt_of_lt_of_lt_two_pow <;> bv_omega
+      rw [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le hdr']
+      omega
+    case neg.hqWidth =>
+      apply toNat_shiftConcat_lt_of_lt_of_lt_two_pow <;> omega
+    case neg.hdiv =>
+      have rltd' := (BitVec.le_iff_not_lt.mp rltd)
+      simp only [qr.toNat_shiftRight_sub_one_eq h]
+      simp only [BitVec.toNat_sub_eq_toNat_sub_toNat_of_le rltd']
+      simp only [toNat_shiftConcat_eq_of_lt_of_lt_two_pow (hk := qr.wr_lt_w h) (hx := h.hrWidth)]
+      simp only [BitVec.le_def, toNat_shiftConcat_eq_of_lt_of_lt_two_pow (hk := qr.wr_lt_w h) (hx := h.hrWidth)] at rltd'
+      simp only [toNat_shiftConcat_eq_of_lt_of_lt_two_pow (hk := qr.wr_lt_w h) (hx := h.hqWidth),
+        h.hdiv, Nat.mul_add]
```

Comment:
Is there a reason this is not a single simp only?

Reply:
Yes, because the interleaved one is ` at rltd'`.